### PR TITLE
Distribution

### LINKED ISSUE #667: Rename `cw20-stake-external-rewards` and refactor to support other voting modules
`cw20-stake-external-rewards` could work with both native token and staked NFT DAOs. #665 and #664 add the hooks needed to make this happen.

- [ ] rename to something more general like `dao-external-staking-rewards`
- [ ] Tests to make sure it works with both NFT and native token staking as well as cw20
- [ ] Refactor to be more readable?

### PR DESCRIPTION
**Pull request by JakeHartnell**
Closes #667 :tada: 

**User Story**: As a DAO, I want to distribute rewards over time to members based on their voting power or stake.

This PR refactors the old `cw20-stake-external-rewards` into something that can be used by _all_ types of DAOs. I've tried my best to add more code comments and documentation into this contract as well, hoping to make it a bit more readable as we move forward.

Additionally, I created a new folder `distribution` as there was starting to be too much in the `external` contracts folder. I moved `cw-fund-distributor` there for now.

As for `cw20-stake-external-rewards`, I have left it where it is for now, but thinking about deleting it...

### CODE REVIEW

File: contracts/distribution/dao-rewards-distributor/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,2175 @@
...
+    );
+    assert_eq!(get_balance_native(&app, ADDR3, &denom), Uint128::new(0));
+    assert_eq!(
+        get_balance_native(&app, &reward_addr, &denom),
+        Uint128::new(24997500)
+    );
+}
+
+#[test]
+fn test_cw4_dao_rewards() {
```
Comment (JakeHartnell) [inline contracts/distribution/dao-rewards-distributor/src/tests.rs:2024]: This test is sadly not working currently, and I'm not quite sure why. :smiling_face_with_tear:

Comment (JakeHartnell) [reply contracts/distribution/dao-rewards-distributor/src/tests.rs:2024]: Fixed! Noticed some interesting behavior around snapshot maps, messages, and submessages.

The fix was to specify the currently block height when querying `VotingPowerAtHeight` and `TotalPowerAtHeight`.

File: Cargo.toml [hunk 2]
```diff
@@ -108,6 +110,7 @@ dao-proposal-hook-counter = { path = "./contracts/test/dao-proposal-hook-counter
 dao-proposal-multiple = { path = "./contracts/proposal/dao-proposal-multiple", version = "2.4.0" }
 dao-proposal-single = { path = "./contracts/proposal/dao-proposal-single", version = "2.4.0" }
 dao-proposal-sudo = { path = "./contracts/test/dao-proposal-sudo", version = "2.4.0" }
+dao-rewards-distributor = { path = "./contracts/distribution/dao-rewards-distributor", version = "*" }
```
Comment (bekauz) [inline Cargo.toml]: is `cw-fund-distributor` and `dao-rewards-distributor` here versioned with wildcard for a reason? also noticed that `cw-fund-distributor` has a hardcoded version of "0.1.0" from before, I pushed a change where it sets it to workspace like we do in other contracts.

File: contracts/distribution/dao-rewards-distributor/src/msg.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    pub vp_contract: String,
+    /// An optional contract that is allowed to call the StakeChangedHook.
+    /// Often, the vp_contract calls hooks for power change events, but sometimes
+    /// they are separate. For example, the cw20-stake contract is separate from
+    /// the dao-voting-cw20-staked contract.
+    pub hook_caller: Option<String>,
+    /// The denom in which rewards are paid out.
+    pub reward_denom: Denom,
+    /// The duration of the reward period in blocks.
+    pub reward_duration: u64,
```
Comment (bekauz) [inline contracts/distribution/dao-rewards-distributor/src/msg.rs:30]: shall we make this work with both block and time based periods?

File: contracts/distribution/dao-rewards-distributor/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+        Cw20(addr) => Cw20(deps.api.addr_validate(addr.as_ref())?),
+    };
+
+    // Verify contract provided is a voting module contract
+    let _: TotalPowerAtHeightResponse = deps.querier.query_wasm_smart(
+        &msg.vp_contract,
+        &VotingQueryMsg::TotalPowerAtHeight { height: None },
+    )?;
+
+    // Optional hook caller is allowed to call voting power change hooks.
```
Comment (bekauz) [inline contracts/distribution/dao-rewards-distributor/src/contract.rs:56]: can you give an example of how this could be used?

**Review by bekauz — COMMENTED** [review]
overall this looks great.
a little hard to review the new code you added as everything appears new because of the directory change, but both `MemberChangedHook` and `NftStakeChangedHook` logic makes sense to me!

File: contracts/distribution/dao-rewards-distributor/src/msg.rs (hunk 3 above)
Comment (noahsaso) [reply contracts/distribution/dao-rewards-distributor/src/msg.rs:30]: yes 🙌

Comment (JakeHartnell) [reply contracts/distribution/dao-rewards-distributor/src/msg.rs:30]: Definitely yes, that would be a huge improvement.

File: Cargo.toml (hunk 2 above)
Comment (JakeHartnell) [reply Cargo.toml]: Yes, it should be workspace version.

File: contracts/distribution/dao-rewards-distributor/src/contract.rs (hunk 4 above)
Comment (JakeHartnell) [reply contracts/distribution/dao-rewards-distributor/src/contract.rs:56]: For example, cw20 DAOs have separate contracts for staking and voting power. Others, depending on their setup, may also have separate contracts for staking and returning voting power.

**Comment by JakeHartnell** [comment]
Note sure why CodeCov is failing, but everything else passes. 🤷‍♂️

**Review by bekauz — APPROVED** [review]

**Comment by bekauz** [comment]
> Note sure why CodeCov is failing, but everything else passes. 🤷‍♂️

passes now!

**Comment by noahsaso** [comment]
closing in favor of #835

**Pull request closed without merging**